scylla-cql/src/types/deserialize/mod.rs

```diff
@@ -211,6 +211,11 @@ impl TypeCheckError {
     pub fn new(err: impl std::error::Error + Send + Sync + 'static) -> Self {
         Self(Arc::new(err))
     }
+
+    /// Retrieve an error reason.
+    pub fn get_reason(&self) -> &Arc<dyn std::error::Error + Send + Sync> {
+        &self.0
+    }
```

Comment:
Do we prefer to expose `&Arc<dyn...>` or `&dyn...`? In the former case, it's possible to clone the underlying error; no idea if it's better or not.

Reply:
With `&dyn` it is still possible to clone the outer error so I don't see the difference.
Downcast methods on `Error` are not defined for `Arc` anyway, only for `Box`, so user will have to use methods on refs anyway.

Reply:
Let's expose bare `&dyn` then.

Reply:
`downcast_ref` method requires `T: 'static`. See: https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref-1. If we return `&dyn Error`, then result will not satisfy `'static` bound. I'm curious why does it work for `Arc<dyn Error>` though - I'd be glad for an explanation.

Reply:
I tried to change `BrokenConnectionError::get_reason` accordingly, but received a compile error (because of the reasons mentioned above).

Reply:
`Arc` is `'static`, because its lifetime is independent of any external data. It manages its data on its own, so it's `'static` just like `Vec` or `String`.

Reply:
> I tried to change `BrokenConnectionError::get_reason` accordingly, but received a compile error (because of the reasons mentioned above).

Forgot to mention that this method is used in one test - compiler error appears there

Reply:
> `downcast_ref` method requires `T: 'static`. See: https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref-1. If we return `&dyn Error`, then result will not satisfy `'static` bound. I'm curious why it works for `Arc<dyn Error>` though - I'd be glad for an explanation.

This is a bound on T, not on the reference. So if you downcast `&dyn Error ` to `SomeError` (so the method returns `Option<&SomeError>` then `SomeError` must be `Send + 'static` - which it typically is. What is the code and error you are getting?

Reply:
> > `downcast_ref` method requires `T: 'static`. See: https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref-1. If we return `&dyn Error`, then result will not satisfy `'static` bound. I'm curious why it works for `Arc<dyn Error>` though - I'd be glad for an explanation.
> 
> This is a bound on T, not on the reference.

Yeah, I just noticed that `T` is a return type. There are no lifetime bounds on arguments

Reply:
Here is a snippet from test:
```rust
        // Wait until keepaliver gots impatient and terminates router.
        // Then, the error from keepaliver will be propagated to the error receiver.
        let err = error_receiver.await.unwrap();
        let err_inner: &BrokenConnectionErrorKind = match err {
            crate::transport::connection::ConnectionError::BrokenConnection(ref e) => {
                e.get_reason().downcast_ref().unwrap()
            }
            _ => panic!("Bad error type. Expected keepalive timeout."),
        };
        assert_matches!(err_inner, BrokenConnectionErrorKind::KeepaliveTimeout(_));
```

It works when `BrokenConnectionError` returns `Arc<dyn ...>`, but when I change the return type to `&dyn Error`, I get the following error:
```
error[E0597]: `err.0` does not live long enough
    --> scylla/src/transport/connection.rs:2782:77
     |
2782 |             crate::transport::connection::ConnectionError::BrokenConnection(ref e) => {
     |                                                                             ^^^^^ borrowed value does not live long enough
2783 |                 e.get_reason().downcast_ref().unwrap()
     |                 -------------- argument requires that `err.0` is borrowed for `'static`
...
2796 |     }
     |     - `err.0` dropped here while still borrowed
```

Reply:
What does it say if you add `drop(err);` after `assert_matches`?

Reply:
> What does it say if you add `drop(err);` after `assert_matches`?

"cannot move out of `err` because it is borrowed"

Reply:
Can you try to assign `e` to a variable first? So the snippet would look like this:
```
        // Wait until keepaliver gots impatient and terminates router.
        // Then, the error from keepaliver will be propagated to the error receiver.
        let err = error_receiver.await.unwrap();
        let err_inner = match err {
            crate::transport::connection::ConnectionError::BrokenConnection(ref e) => {
                e
            }
            _ => panic!("Bad error type. Expected keepalive timeout."),
        };
        let err_typed: &BrokenConnectionErrorKind = err_inner.get_reason().downcast_ref().unwrap();
        assert_matches!(err_typed, BrokenConnectionErrorKind::KeepaliveTimeout(_));
```

Reply:
The same error:
```
error[E0597]: `err.0` does not live long enough
    --> scylla/src/transport/connection.rs:2782:77
     |
2782 |             crate::transport::connection::ConnectionError::BrokenConnection(ref e) =...
     |                                                                             ^^^^^ borrowed value does not live long enough
...
2785 |         let err_typed: &BrokenConnectionErrorKind = err_inner.get_reason().downcast_...
     |                                                     ---------------------- argument requires that `err.0` is borrowed for `'static`
...
2795 |     }
     |     - `err.0` dropped here while still borrowed
```

Reply:
Wtf. 
What docs.rs show:
![image](https://github.com/user-attachments/assets/f5e57cbb-1f46-4af6-8c2b-7fc36045bb15)
What is really there when you click "source":
![Uploading image.png…]()

Reply:
Wtf. 
What docs.rs show:
![image](https://github.com/user-attachments/assets/f5e57cbb-1f46-4af6-8c2b-7fc36045bb15)
What is really there when you click "source":
![image](https://github.com/user-attachments/assets/d2aefef7-7e97-4d59-afd1-3d5a4e4e28df)

Reply:
In that case, I think the best solution is to create our own downcast_ref method on TypeCheckError etc - that should be possible because inside we have access to Arc, right?

Reply:
I'll try it out. But I still don't really get it why it works for `Arc<dyn Error>`, while it does not for `&dyn Error`. AFAIU, what happens when we call `downcast_ref` on `Arc<dyn Error>` is value inside `Arc` is implicitly dereferenced first (i.e. `&dyn Error` is retrieved automatically via `Deref` trait). Then, `downcast_ref` is called on obtained `&dyn Error` reference. It seems almost identical to me.

Reply:
We should understand it and decide on how to best fix that before merging. I'll try to look into this today or during the weekend.